### ISSUE TITLE
Unable to install (`error: command 'gcc' failed with exit status 1`)

### ISSUE THREAD

**Issue by abeburnett** [ing-bank/sparse_dot_topn#7]
I've been trying everything to get this to successfully install. I upgraded my MacOS to the most recent edition, updated XCode and XCode Dev Tools. Did `brew doctor` and `brew upgrade`. Upgraded Python from 3.5 --> 3.7. No dice. I continue to get this error:
```

python3 -m pip install git+https://github.com/ing-bank/sparse_dot_topn
Collecting git+https://github.com/ing-bank/sparse_dot_topn
  Cloning https://github.com/ing-bank/sparse_dot_topn to /private/var/folders/p9/shlr4jjn61b5ts3bbsz7wkym0000gn/T/pip-req-build-3u5wczkm
Building wheels for collected packages: sparse-dot-topn
  Building wheel for sparse-dot-topn (setup.py) ... error
  Complete output from command /Users/Abe/miniconda3/bin/python3 -u -c "import setuptools, tokenize;__file__='/private/var/folders/p9/shlr4jjn61b5ts3bbsz7wkym0000gn/T/pip-req-build-3u5wczkm/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /private/var/folders/p9/shlr4jjn61b5ts3bbsz7wkym0000gn/T/pip-wheel-plbrmov8 --python-tag cp37:
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.macosx-10.7-x86_64-3.7
  creating build/lib.macosx-10.7-x86_64-3.7/sparse_dot_topn
  copying sparse_dot_topn/__init__.py -> build/lib.macosx-10.7-x86_64-3.7/sparse_dot_topn
  copying sparse_dot_topn/awesome_cossim_topn.py -> build/lib.macosx-10.7-x86_64-3.7/sparse_dot_topn
  running build_ext
  building 'sparse_dot_topn.sparse_dot_topn' extension
  creating build/temp.macosx-10.7-x86_64-3.7
  creating build/temp.macosx-10.7-x86_64-3.7/sparse_dot_topn
  gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/Abe/miniconda3/include -arch x86_64 -I/Users/Abe/miniconda3/include -arch x86_64 -I./sparse_dot_topn -I/Users/Abe/miniconda3/lib/python3.7/site-packages/numpy/core/include -I/Users/Abe/miniconda3/include/python3.7m -c ./sparse_dot_topn/sparse_dot_topn.cpp -o build/temp.macosx-10.7-x86_64-3.7/./sparse_dot_topn/sparse_dot_topn.o -std=c++0x -Os
  warning: include path for stdlibc++ headers not found; pass '-stdlib=libc++' on the command line to use the libc++ standard library instead [-Wstdlibcxx-not-found]
  In file included from ./sparse_dot_topn/sparse_dot_topn.cpp:634:
  In file included from /Users/Abe/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4:
  In file included from /Users/Abe/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:12:
  In file included from /Users/Abe/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1822:
  /Users/Abe/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: "Using deprecated NumPy API, disable it with "          "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-W#warnings]
  #warning "Using deprecated NumPy API, disable it with " \
   ^
  ./sparse_dot_topn/sparse_dot_topn.cpp:670:14: fatal error: 'cstdlib' file not found
      #include <cstdlib>
               ^~~~~~~~~
  2 warnings and 1 error generated.
  error: command 'gcc' failed with exit status 1

  ----------------------------------------
  Failed building wheel for sparse-dot-topn
  Running setup.py clean for sparse-dot-topn
Failed to build sparse-dot-topn
Installing collected packages: sparse-dot-topn
  Running setup.py install for sparse-dot-topn ... error
    Complete output from command /Users/Abe/miniconda3/bin/python3 -u -c "import setuptools, tokenize;__file__='/private/var/folders/p9/shlr4jjn61b5ts3bbsz7wkym0000gn/T/pip-req-build-3u5wczkm/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/var/folders/p9/shlr4jjn61b5ts3bbsz7wkym0000gn/T/pip-record-z7l2po1a/install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build/lib.macosx-10.7-x86_64-3.7
    creating build/lib.macosx-10.7-x86_64-3.7/sparse_dot_topn
    copying sparse_dot_topn/__init__.py -> build/lib.macosx-10.7-x86_64-3.7/sparse_dot_topn
    copying sparse_dot_topn/awesome_cossim_topn.py -> build/lib.macosx-10.7-x86_64-3.7/sparse_dot_topn
    running build_ext
    building 'sparse_dot_topn.sparse_dot_topn' extension
    creating build/temp.macosx-10.7-x86_64-3.7
    creating build/temp.macosx-10.7-x86_64-3.7/sparse_dot_topn
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/Abe/miniconda3/include -arch x86_64 -I/Users/Abe/miniconda3/include -arch x86_64 -I./sparse_dot_topn -I/Users/Abe/miniconda3/lib/python3.7/site-packages/numpy/core/include -I/Users/Abe/miniconda3/include/python3.7m -c ./sparse_dot_topn/sparse_dot_topn.cpp -o build/temp.macosx-10.7-x86_64-3.7/./sparse_dot_topn/sparse_dot_topn.o -std=c++0x -Os
    warning: include path for stdlibc++ headers not found; pass '-stdlib=libc++' on the command line to use the libc++ standard library instead [-Wstdlibcxx-not-found]
    In file included from ./sparse_dot_topn/sparse_dot_topn.cpp:634:
    In file included from /Users/Abe/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4:
    In file included from /Users/Abe/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:12:
    In file included from /Users/Abe/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1822:
    /Users/Abe/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: "Using deprecated NumPy API, disable it with "          "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-W#warnings]
    #warning "Using deprecated NumPy API, disable it with " \
     ^
    ./sparse_dot_topn/sparse_dot_topn.cpp:670:14: fatal error: 'cstdlib' file not found
        #include <cstdlib>
                 ^~~~~~~~~
    2 warnings and 1 error generated.
    error: command 'gcc' failed with exit status 1

    ----------------------------------------
Command "/Users/Abe/miniconda3/bin/python3 -u -c "import setuptools, tokenize;__file__='/private/var/folders/p9/shlr4jjn61b5ts3bbsz7wkym0000gn/T/pip-req-build-3u5wczkm/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/var/folders/p9/shlr4jjn61b5ts3bbsz7wkym0000gn/T/pip-record-z7l2po1a/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /private/var/folders/p9/shlr4jjn61b5ts3bbsz7wkym0000gn/T/pip-req-build-3u5wczkm/
```

Any ideas? I tried the .whl but it appears to be for Python 3.6 only. Please advise!

**Comment by abeburnett** [comment]
So far anyone else on a Mac who wanders across my issue, I seem to have resolved it by adding to my `.bash_profile`:

export CC=/usr/local/bin/gcc-8
export CXX=/usr/local/bin/g++-8

And doing `source ~/.bash_profile` afterwards. 

For whatever reason, the issue seems to be that the `extra_compile_args` in the `setup.py` are setting up the C compiler incorrectly for the standard Mac XCode tooling. Either that, or the Mac compiler setup is incompatible. Either way, forcing the usage of GCC (8.3.0) in my case seems to allowed the installation to proceed correctly. 

Oh, and I did add to the `setup.py`:

```
import platform

if platform.system() == "Darwin":
    compile_extra_args = ['-std=c++11', "-mmacosx-version-min=10.14"]
```

And change the reference to `compile_extra_args` here:

```
ext_utils = Extension('sparse_dot_topn.sparse_dot_topn',
                      sources=['./sparse_dot_topn/sparse_dot_topn.pyx', './sparse_do$
                      include_dirs=[numpy.get_include()],
                      #libraries=[],
                      extra_compile_args=compile_extra_args,
                      language='c++',
                     )
```

Hope this helps someone!

**Comment by ymwdalex** [comment]
@abeburnett thanks for your solution!

**Comment by nicornk** [comment]
@abeburnett thanks for the nice solution!

**Comment by qinyufm** [comment]
have the same issue but on windows.

**Comment by ymwdalex** [comment]
@qinyufm someone suggests use `conda install -c conda-forge sparse_dot_topn`. Could you please give a try to see if it work?

**Comment by qinyufm** [comment]
I don't have conda, only pip install working on my current windows. Is there similar command but using pip?

**Comment by ymwdalex** [comment]
@qinyufm could you please follow this thread https://github.com/ing-bank/sparse_dot_topn/issues/39 to see if it works?

**Comment by avynychenko** [comment]
> @qinyufm someone suggests use `conda install -c conda-forge sparse_dot_topn`. Could you please give a try to see if it work?

Thanks for your help!
It allowed me to resolve problem with installation

**Comment by samitshah** [comment]
The following worked for me:

`CFLAGS="-stdlib=libc++" CXXFLAGS="-stdlib=libc++" MACOSX_DEPLOYMENT_TARGET="10.10" pip install sparse_dot_topn`

**Comment by giacomo-ensuro** [comment]
Hi I am having trouble with using this package (not necessarily with installing it). 
When importing it I get this error (see attached) which I tried to solve in many ways (installing uninstalling numpy, cython, etcetc). 

Do you have any suggestions? 

<img width="919" alt="Screenshot 2022-04-12 at 15 59 56" src="https://user-images.githubusercontent.com/99185234/163004869-b1a0d4be-4e41-47de-bef4-3eba93089000.png">

**Comment by dbejaranoch** [comment]
Hi there! I'm seeing the same 'gcc' issue but on an AWS instance. The OS is "Amazon Linux AMI 2018.03". 

Does anyone know if there's a version of CFLAGS and CXXFLAGS variable configurations for such OS. Similar to:
`CFLAGS="-stdlib=libc++" CXXFLAGS="-stdlib=libc++" MACOSX_DEPLOYMENT_TARGET="10.10" pip install sparse_dot_topn`

**Comment by ymwdalex** [comment]
@dbejaranoch @giacomo-ensuro @avynychenko @qinyufm @samitshah @nicornk @abeburnett 

Thanks for involving the discussion of this thread! We just released [v0.3.3](https://pypi.org/project/sparse-dot-topn/#files). We employed Github actions and CiBuildWheel, and make the wheels available for Py3.6-3.10 on multiple architectures of Linux, MacOS and Windows.

If you still have installation problems, please try the latest version.

If you find any problem with v0.3.3, please let us know. Thanks!

**Comment by ymwdalex** [comment]
[v0.3.3](https://pypi.org/project/sparse-dot-topn/#files) should fix this issue
